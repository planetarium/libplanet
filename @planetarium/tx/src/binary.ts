
export function compareUint8Array(a: Uint8Array, b: Uint8Array) {
  const length = Math.min(a.length, b.length);
  for (let i = 0; i < length; i++) {
    if (a[i] !== b[i]) return a[i] - b[i];
  }
  return a.length - b.length;
}
