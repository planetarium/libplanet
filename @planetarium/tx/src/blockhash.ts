import { Encodable } from "bencodex";

export type BlockHash = Uint8Array;  // TODO: proper type definition

export function encodeBlockHash(blockHash: BlockHash): Encodable {
  if (blockHash.length !== 32) {
    throw new TypeError(`BlockHash must be 32 bytes long, but got ${
      blockHash.length} bytes.`);
  }
  return blockHash.buffer;
}
