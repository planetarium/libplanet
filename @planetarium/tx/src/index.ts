export { type Address, encodeAddress } from "./address";
export {
  type Currency,
  encodeCurrency,
  type FungibleAssetValue,
  getCurrencyHash,
  getMajorUnit,
  getMinorUnit,
  getSign,
} from "./assets";
export {
  encodeMint,
  encodeSystemAction,
  encodeTransfer,
  type Mint,
  type SystemAction,
  type Transfer,
} from "./action";
export type { TxMetadata, UnsignedTxWithSystemAction } from "./tx";
export { encodeTxMetadata, encodeUnsignedTxWithSystemAction } from "./tx";
