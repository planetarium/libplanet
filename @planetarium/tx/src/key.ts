import { Encodable } from "bencodex";

export type PublicKey = Uint8Array;  // TODO: proper type definition

export function encodePublicKey(publicKey: PublicKey): Encodable {
  if (publicKey.length < 1) {
    throw new TypeError("Public key must not be empty.");
  }
  return publicKey.buffer;
}
