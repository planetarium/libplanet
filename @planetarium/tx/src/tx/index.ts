export type { TxMetadata } from "./metadata";
export { encodeTxMetadata } from "./metadata";
export type { UnsignedTxWithSystemAction } from "./unsigned";
export { encodeUnsignedTxWithSystemAction } from "./unsigned";
