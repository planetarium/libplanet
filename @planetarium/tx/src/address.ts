import { Encodable } from "bencodex";
import { compareUint8Array } from "./binary";

export type Address = Uint8Array;  // TODO: proper type definition

export function encodeAddress(address: Address): Encodable {
  if (address.length !== 20) {
    throw new TypeError(`Address must be 20 bytes long, but got ${
      address.length} bytes.`);
  }
  return address.buffer;
}

export function encodeAddressSet(addresses: Set<Address>): Encodable {
  const array: Address[] = [];
  addresses.forEach(addr => array.push(addr));
  array.sort(compareUint8Array);
  return array.map(encodeAddress);
}
