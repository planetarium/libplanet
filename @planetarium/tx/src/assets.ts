import { Encodable, encode } from "bencodex";
import { Address, encodeAddressSet } from "./address";

export interface Currency {
  ticker: string;
  decimalPlaces: number;
  minters: Set<Address> | null;
  totalSupplyTrackable: boolean;
  maximumSupply: {
    major: bigint;
    minor: bigint;
  } | null;
}

export function encodeCurrency(currency: Currency): Encodable {
  const minters: Encodable = currency.minters === null
    ? null
    : encodeAddressSet(currency.minters);
  const serialized: Encodable = {
    ticker: currency.ticker,
    decimals: currency.decimalPlaces,
    minters,
  };

  if (currency.maximumSupply !== null) {
    if (!currency.totalSupplyTrackable) {
      throw new TypeError("maximumSupply implies totalSupplyTrackable");
    }

    serialized.maximumSupplyMajor = currency.maximumSupply.major;
    serialized.maximumSupplyMinor = currency.maximumSupply.minor;
  }

  if (currency.totalSupplyTrackable) {
    serialized.totalSupplyTrackable = true;
  }

  return serialized;
}

export async function getCurrencyHash(currency: Currency): Promise<Uint8Array> {
  const encoded = encode(encodeCurrency(currency));
  const buffer = await crypto.subtle.digest("SHA-1", encoded);
  return new Uint8Array(buffer);
}

export interface FungibleAssetValue {
  rawValue: bigint;
  currency: Currency;
}

function abs(value: bigint): bigint {
  return value < 0n ? -value : value;
}

export function getSign(value: FungibleAssetValue): -1 | 0 | 1 {
  return value.rawValue < 0n ? -1 : value.rawValue > 0n ? 1 : 0;
}

export function getMajorUnit(value: FungibleAssetValue): bigint {
  return abs(value.rawValue) / 10n ** BigInt(value.currency.decimalPlaces);
}

export function getMinorUnit(value: FungibleAssetValue): bigint {
  return abs(value.rawValue) % 10n ** BigInt(value.currency.decimalPlaces);
}
