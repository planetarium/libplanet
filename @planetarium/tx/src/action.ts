import { Encodable } from "bencodex";
import { Address, encodeAddress } from "./address";
import { encodeCurrency, FungibleAssetValue } from "./assets";

export type SystemAction = Mint | Transfer;

export function encodeSystemAction(action: SystemAction): Encodable {
  switch (action.type) {
  case "mint":
    return {
      type_id: 0,
      values: encodeMint(action),
    };
  case "transfer":
    return {
      type_id: 1,
      values: encodeTransfer(action),
    };
  default:
    throw new TypeError(`Unknown action type: ${action["type"]}`);
  }
}

export interface Mint {
  type: "mint";
  recipient: Address;
  amount: FungibleAssetValue;
}

export function encodeMint(action: Mint): Encodable {
  return {
    recipient: encodeAddress(action.recipient),
    currency: encodeCurrency(action.amount.currency),
    amount: action.amount.rawValue,
  };
}

export interface Transfer {
  type: "transfer";
  recipient: Address;
  amount: FungibleAssetValue;
}

export function encodeTransfer(action: Transfer): Encodable {
  return {
    recipient: encodeAddress(action.recipient),
    currency: encodeCurrency(action.amount.currency),
    amount: action.amount.rawValue,
  };
}
