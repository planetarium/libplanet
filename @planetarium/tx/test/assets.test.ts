import { encode } from "bencodex";
import { expect, test } from "vitest";
import { Address } from "../src/address";
import {
  Currency,
  encodeCurrency,
  getCurrencyHash,
  getMajorUnit,
  getMinorUnit,
  getSign
} from "../src/assets";
import { fromHex, toHex } from "./hex";

if (globalThis.crypto == null) {
  // FIXME: This is a workaround for the lack of Web Crypto API in Vitest.
  globalThis.crypto = require("node:crypto");
}

const addressA: Address = fromHex("D6D639DA5a58A78A564C2cD3DB55FA7CeBE244A9");
const addressB: Address = fromHex("5003712B63baAB98094aD678EA2B24BcE445D076");

test("encodeCurrency", () => {
  const encoded = encodeCurrency({
    ticker: "FOO",
    decimalPlaces: 2,
    minters: null,
    totalSupplyTrackable: false,
    maximumSupply: null,
  });
  expect(encoded).toMatchSnapshot();
  expect(encode(encoded)).toMatchSnapshot();

  const encoded2 = encodeCurrency({
    ticker: "BAR",
    decimalPlaces: 0,
    minters: new Set([addressA, addressB]),
    totalSupplyTrackable: true,
    maximumSupply: { major: 100n, minor: 0n },
  });
  expect(encoded2).toMatchSnapshot();
  expect(encode(encoded2)).toMatchSnapshot();
});

test("getCurrencyHash", async () => {
  const legacy = await getCurrencyHash({
    ticker: "GOLD",
    decimalPlaces: 2,
    minters: new Set([addressA]),
    totalSupplyTrackable: false,
    maximumSupply: null,
  });
  expect(toHex(legacy)).toMatchSnapshot();

  const legacy2 = await getCurrencyHash({
    ticker: "NCG",
    decimalPlaces: 8,
    minters: new Set([addressA, addressB]),
    totalSupplyTrackable: false,
    maximumSupply: null,
  });
  expect(toHex(legacy2)).toMatchSnapshot();

  const legacy3 = await getCurrencyHash({
    ticker: "FOO",
    decimalPlaces: 0,
    minters: new Set([]),
    totalSupplyTrackable: false,
    maximumSupply: null,
  });
  expect(toHex(legacy3)).toMatchSnapshot();

  const legacy4 = await getCurrencyHash({
    ticker: "BAR",
    decimalPlaces: 1,
    minters: null,
    totalSupplyTrackable: false,
    maximumSupply: null,
  });
  expect(toHex(legacy4)).toMatchSnapshot();

  const uncapped = await getCurrencyHash({
    ticker: "BAZ",
    decimalPlaces: 1,
    minters: null,
    totalSupplyTrackable: true,
    maximumSupply: null,
  });
  expect(toHex(uncapped)).toMatchSnapshot();

  const capped = await getCurrencyHash({
    ticker: "BAZ",
    decimalPlaces: 1,
    minters: null,
    totalSupplyTrackable: true,
    maximumSupply: { major: 100n, minor: 0n },
  });
  expect(toHex(capped)).toMatchSnapshot();
});

const FOO: Currency = {
  ticker: "FOO",
  decimalPlaces: 2,
  minters: null,
  totalSupplyTrackable: true,
  maximumSupply: null,
};

const fooA = { rawValue: 12345n, currency: FOO };
const fooB = { rawValue: 45609n, currency: FOO };
const fooC = { rawValue: 10n, currency: FOO };
const fooD = { rawValue: 9n, currency: FOO };
const fooE = { rawValue: -78901n, currency: FOO };
const fooF = { rawValue: -2n, currency: FOO };
const fooG = { rawValue: 12300n, currency: FOO };
const fooH = { rawValue: 0n, currency: FOO };

test("getSign", () => {
  expect(getSign(fooA)).toBe(1);
  expect(getSign(fooB)).toBe(1);
  expect(getSign(fooC)).toBe(1);
  expect(getSign(fooD)).toBe(1);
  expect(getSign(fooE)).toBe(-1);
  expect(getSign(fooF)).toBe(-1);
  expect(getSign(fooG)).toBe(1);
  expect(getSign(fooH)).toBe(0);
});

test("getMajorUnit", () => {
  expect(getMajorUnit(fooA)).toBe(123n);
  expect(getMajorUnit(fooB)).toBe(456n);
  expect(getMajorUnit(fooC)).toBe(0n);
  expect(getMajorUnit(fooD)).toBe(0n);
  expect(getMajorUnit(fooE)).toBe(789n);
  expect(getMajorUnit(fooF)).toBe(0n);
  expect(getMajorUnit(fooG)).toBe(123n);
  expect(getMajorUnit(fooH)).toBe(0n);
});

test("getMinorUnit", () => {
  expect(getMinorUnit(fooA)).toBe(45n);
  expect(getMinorUnit(fooB)).toBe(9n);
  expect(getMinorUnit(fooC)).toBe(10n);
  expect(getMinorUnit(fooD)).toBe(9n);
  expect(getMinorUnit(fooE)).toBe(1n);
  expect(getMinorUnit(fooF)).toBe(2n);
  expect(getMinorUnit(fooG)).toBe(0n);
  expect(getMinorUnit(fooH)).toBe(0n);
});
