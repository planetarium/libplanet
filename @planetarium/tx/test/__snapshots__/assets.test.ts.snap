// Vitest Snapshot v1

exports[`encodeCurrency 1`] = `
{
  "decimals": 2,
  "minters": null,
  "ticker": "FOO",
}
`;

exports[`encodeCurrency 2`] = `
{
  "data": [
    100,
    117,
    56,
    58,
    100,
    101,
    99,
    105,
    109,
    97,
    108,
    115,
    105,
    50,
    101,
    117,
    55,
    58,
    109,
    105,
    110,
    116,
    101,
    114,
    115,
    110,
    117,
    54,
    58,
    116,
    105,
    99,
    107,
    101,
    114,
    117,
    51,
    58,
    70,
    79,
    79,
    101,
  ],
  "type": "Buffer",
}
`;

exports[`encodeCurrency 3`] = `
{
  "decimals": 0,
  "maximumSupplyMajor": 100n,
  "maximumSupplyMinor": 0n,
  "minters": [
    ArrayBuffer [],
    ArrayBuffer [],
  ],
  "ticker": "BAR",
  "totalSupplyTrackable": true,
}
`;

exports[`encodeCurrency 4`] = `
{
  "data": [
    100,
    117,
    56,
    58,
    100,
    101,
    99,
    105,
    109,
    97,
    108,
    115,
    105,
    48,
    101,
    117,
    49,
    56,
    58,
    109,
    97,
    120,
    105,
    109,
    117,
    109,
    83,
    117,
    112,
    112,
    108,
    121,
    77,
    97,
    106,
    111,
    114,
    105,
    49,
    48,
    48,
    101,
    117,
    49,
    56,
    58,
    109,
    97,
    120,
    105,
    109,
    117,
    109,
    83,
    117,
    112,
    112,
    108,
    121,
    77,
    105,
    110,
    111,
    114,
    105,
    48,
    101,
    117,
    55,
    58,
    109,
    105,
    110,
    116,
    101,
    114,
    115,
    108,
    50,
    48,
    58,
    80,
    3,
    113,
    43,
    99,
    186,
    171,
    152,
    9,
    74,
    214,
    120,
    234,
    43,
    36,
    188,
    228,
    69,
    208,
    118,
    50,
    48,
    58,
    214,
    214,
    57,
    218,
    90,
    88,
    167,
    138,
    86,
    76,
    44,
    211,
    219,
    85,
    250,
    124,
    235,
    226,
    68,
    169,
    101,
    117,
    54,
    58,
    116,
    105,
    99,
    107,
    101,
    114,
    117,
    51,
    58,
    66,
    65,
    82,
    117,
    50,
    48,
    58,
    116,
    111,
    116,
    97,
    108,
    83,
    117,
    112,
    112,
    108,
    121,
    84,
    114,
    97,
    99,
    107,
    97,
    98,
    108,
    101,
    116,
    101,
  ],
  "type": "Buffer",
}
`;

exports[`getCurrencyHash 1`] = `"81446cd346c1be9e686835742bfd3772194dea21"`;

exports[`getCurrencyHash 2`] = `"42ce3a098fe14084e89d3d4449f56126693aeed1"`;

exports[`getCurrencyHash 3`] = `"801990ea2885bd51eebca0e826cc0e27f0917a9b"`;

exports[`getCurrencyHash 4`] = `"da42781871890f1e1b7d6f49c7f2733d3ba7b8bd"`;

exports[`getCurrencyHash 5`] = `"d7fe111cae5b2503939c9bce864ca3b64d575e8d"`;

exports[`getCurrencyHash 6`] = `"38bd85ea71c09ca7ed82b61fe91bc205101db191"`;
